src/hooks/useAsyncActionHandler/useAsyncActionHandler.ts

```diff
@@ -0,0 +1,48 @@
+import React from 'react';
+
+export interface UseAsyncActionHandlerProps<ActionResult = void> {
+    handler: (...args: any[]) => Promise<ActionResult>;
+}
+
+export interface UseAsyncActionHandlerResult<Action = () => Promise<void>> {
+    isLoading: boolean;
+    handler: Action;
+}
+
+export function useAsyncActionHandler<ActionResult>({
+    handler,
+}: UseAsyncActionHandlerProps<ActionResult>): UseAsyncActionHandlerResult<typeof handler> {
+    const [isLoading, setLoading] = React.useState(false);
+
+    const handleAction: typeof handler = React.useCallback(
+        (...args) => {
+            setLoading(true);
+
+            const handlerResult = handler(...args);
+
+            if (handlerResult instanceof Promise) {
+                return handlerResult.finally(() => {
+                    setLoading(false);
+                });
+            }
+
+            if (process.env.NODE_ENV !== 'production') {
+                // used only in dev build
+                // eslint-disable-next-line no-console
+                console.error(
+                    "@gravity-ui/uikit: (useAsyncActionHandler) the provided handler doesn't return a promise",
+                );
+            }
+
+            setLoading(false);
+
+            return handlerResult;
+        },
```

Comment:
```suggestion
        async (...args) => {
            setLoading(true);

            try {
                return await handler(...args);
            } finally {
                setLoading(false);
            }
        },
```

Reply:
1) We should notify the user that the handler is not a promise, otherwise the use of the hook will not be justified 
2) The error should be caught by the user, not by us

Reply:
1. Why? The types say we are waiting for a PromiseLike and if the user passes anything else they won't see the loader.
2. We do not catch errors here, they will be propagated to the user.

Reply:
The user of the library can use `as`